### PR TITLE
Make sick haptic feedback on Android

### PR DESCRIPTION
**Pull request by osdnk**
Fixes RNBW-4020
Figma link (if any):

## What changed (plus any additional context for devs)
I created a matching implementation of feedback that is the same on Android as on iOS. I used the newer API so around 10% of devices will be not covered, but this is ok!

 
Some phones (e.g, my OnePlus 7) are not able to make such a short vibration and they need to be very subtle. Then they are just not working and this is OK. Better to now have them than have them too long. 

## Screen recordings / screenshots
no recording, try it. No impact on iOS
## What to test
1. check buttons (selection haptic)
2. copy address (light impact)
3. touch charts (heavy impact)
4. scan QR successfully (notification success)
5. scan unsupported QA (notification error)
6. Press  "Insufficient funds" (notification warning)

They should not be exactly the same. Just soft and expressing more or less what we need 

Please test on some new and popular phones, e.g., samsung s21

### CODE REVIEW

**Comment by tchayen** [comment]
Is '10%' a made up number?

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+class RNHapticsPerformer {
+  private final Vibrator mVibrator;
+  RNHapticsPerformer(Vibrator vibrator) {
+    mVibrator = vibrator;
+  }
+
+  void notificationSuccess() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      mVibrator.vibrate(VibrationEffect.createWaveform(new long[]{20, 65, 21}, new int[]{145, 0, 130}, -1));
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:16]: Where are those numbers coming from? Do we care about potentially editing them in future?

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import android.os.Vibrator;
+
+class RNHapticsPerformer {
+  private final Vibrator mVibrator;
+  RNHapticsPerformer(Vibrator vibrator) {
+    mVibrator = vibrator;
+  }
+
+  void notificationSuccess() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:15]: What is `.O`?

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      mVibrator.vibrate(VibrationEffect.createWaveform(new long[]{4}, new int[]{30}, -1));
+    }
+  }
+
+  void selection() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      mVibrator.vibrate(VibrationEffect.createWaveform(new long[]{2}, new int[]{60}, -1));
+    }
+  }
+}
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:50]:
```suggestion
}
```

Comment (tchayen) [reply android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:50]: I mean missing new line in the end of file. Not sure if this does what I think it does.

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPackage.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    public List<Class<? extends JavaScriptModule>> createJSModules() {
+      return Collections.emptyList();
+    }
+
+    @Override
+    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
+      return Collections.emptyList();
+    }
+}
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPackage.java:27]: New line

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  }
+
+  @ReactMethod
+  public void selection(){
+    mPerformer.selection();
+  }
+
+  @ReactMethod
+  public void impact(String type){
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java:46]:
```suggestion
  public void impact(String type) {
```

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        mPerformer.notificationWarning();
+        break;
+      default:
+        mPerformer.notificationSuccess();
+        break;
+    }
+  }
+
+  @ReactMethod
+  public void selection(){
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java:41]:
```suggestion
  public void selection() {
```

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    mPerformer = new RNHapticsPerformer(vibrator);
+  }
+
+  @Override
+  public String getName() {
+      return "RNRainbowHaptics";
+  }
+
+  @ReactMethod
+  public void notification(String type){
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java:26]:
```suggestion
  public void notification(String type) {
```

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    switch (type) {
+      case "light":
+        mPerformer.impactLight();
+        break;
+      case "heavy":
+        mPerformer.impactHeavy();
+        break;
+    }
+  }
+}
```
Comment (tchayen) [inline android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNRainbowHapticsModule.java:56]: New line

File: ios/Podfile.lock [hunk 9]
```diff
@@ -1123,4 +1123,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: b28f879725507f95d94adc57febf42531d6e72d1
 
-COCOAPODS: 1.11.2
+COCOAPODS: 1.11.3
```
Comment (tchayen) [inline ios/Podfile.lock:1126]: Is it related?

File: patches/react-native-haptic-feedback+1.11.0.patch [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+     enableVibrateFallback: false,
+@@ -10,7 +10,30 @@ class RNReactNativeHapticFeedback {
+         const triggerOptions = createTriggerOptions(options)
+ 
+         try {
+-            NativeModules.RNReactNativeHapticFeedback.trigger(type, triggerOptions);
++            if (Platform.OS === "android") {
++                switch (type) {
++                    case 'impactLight':
++                        NativeModules.RNRainbowHaptics.impact("light")
```
Comment (tchayen) [inline patches/react-native-haptic-feedback+1.11.0.patch:19]: Do we care about `;` here?

File: src/components/value-chart/Chart.js [hunk 11]
```diff
@@ -202,7 +202,7 @@ export default function ChartWrapper({
             <ChartPath
               fill="none"
               gestureEnabled={!fetchingCharts && !!throttledData}
-              hapticsEnabled={ios}
+              hapticsEnabled
```
Comment (tchayen) [inline src/components/value-chart/Chart.js:205]: Maybe it can now be default?

**Review by tchayen — APPROVED** [review]
Great job!

Leaving some not very useful code style comments.

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java (hunk 1 above)
Comment (osdnk) [reply android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:16]: my intuition

File: android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java (hunk 2 above)
Comment (osdnk) [reply android/app/src/main/java/me/rainbow/NativeModules/Haptics/RNHapticsPerformer.java:15]: version code of android? what?

File: ios/Podfile.lock (hunk 9 above)
Comment (osdnk) [reply ios/Podfile.lock:1126]: no

File: patches/react-native-haptic-feedback+1.11.0.patch (hunk 10 above)
Comment (osdnk) [reply patches/react-native-haptic-feedback+1.11.0.patch:19]: no

**Review by Kemalsirajo — APPROVED** [review]

**Comment by osdnk** [comment]
@tchayen this is not made-up number https://www.appbrain.com/stats/top-android-sdk-versions

**Review by brunobar79 — APPROVED** [review]
NICE

**Comment by BrodyHughes** [comment]
This seems to be functioning fine for me. Can't find anything wrong here, what phones did u test with? The normal haptic seems a little intense for me on my Pixel 6.

**Review by BrodyHughes — APPROVED** [review]
🚀

**Review by estebanmino — APPROVED** [review]
sick

**Review by terrysahaidak — APPROVED** [review]

**Pull request merged**